rust/src/io.rs

```diff
@@ -79,21 +89,141 @@ impl<R: Read + Seek> FileReader<R> {
     pub fn new(file: R) -> Result<Self> {
         let mut f = file;
         let metadata_pos = read_footer(&mut f)?;
-        let metadata: crate::format::pb::Metadata = ProtoParser::read(&mut f, metadata_pos)?;
+        let pb_meta: crate::format::pb::Metadata = ProtoParser::read(&mut f, metadata_pos)?;
         let manifest: crate::format::pb::Manifest =
-            ProtoParser::read(&mut f, metadata.manifest_position as i64)?;
+            ProtoParser::read(&mut f, pb_meta.manifest_position as i64)?;
+        let num_columns = manifest.fields.len();
+        let num_batches = pb_meta.batch_offsets.len() - 1;
+        let page_table = PageTable::make(
+            &mut f,
+            pb_meta.page_table_position,
+            num_columns,
+            num_batches,
+        );
+        let metadata = Metadata::make(pb_meta);
         Ok(FileReader {
             file: f,
             schema: Schema::from_proto(&manifest.fields),
             metadata,
+            page_table,
         })
     }
 
     pub fn schema(&self) -> &Schema {
         &self.schema
     }
 
-    pub fn num_chunks(&self) -> i32 {
-        self.metadata.batch_offsets.len() as i32 - 1
+    pub fn num_chunks(&self) -> usize {
+        self.metadata.num_chunks()
+    }
+
+    pub fn get(&mut self, idx: u32) -> Vec<Box<dyn Scalar>> {
+        // TODO usize for idiomatic ? or i32 for protobuf compatibility?
```

Comment:
I'd go for `rust` idiomatic. 

Btw, is `get / get_array / get_list_array` the idiomatic interfaces in rust? I assume you are trying to mimic the C++ interface? 

I have not checked out `arrow2` too much. What is their naming convention?

Reply:
I was asking about signed integer/unsigned integer conversions here, which is also discussed in [another thread](https://github.com/eto-ai/lance/pull/274#discussion_r1020856448). For `get / get_array / get_list_array`, I think it's a lance thing, not `rust` or `arrow2` thing, we can decide it with out align to `rust` or `arrow2`

Reply:
Through my limited understanding, the name conventions of Arrow C++ / Arrow-rs / Arrow2 are very close. I guess they have considerations on easy migration or quick match-up.